### DISCUSSION TITLE
Question: UWP WinUI2 and Webview2 Fixed distribution on Windows Team (Surface Hub)

### DISCUSSION THREAD

**Discussion by alexispolakMURAL** [microsoft/microsoft-ui-xaml#7439] [Q&A]
I am having trouble finding if it is possible to do this.

I have a UWP WinUI2 app that uses WebView2 and I need the app to work in Windows Teams (a version of windows that  Surface Hub devices have) but on these devices the WebView2 Runtime cannot be installed because only the microsoft store can install apps.

I would like to be able to include the WebView2 runtime binaries inside my application so that an external installation is not necessary. The problem is that most of the documentation or guides are for WPF applications and the process is different is not clear. For example, points 3, 4 and 5 of the "Details about the Fixed Version runtime distribution mode" in this documentation https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution aren't clear enough.

**Comment by asklar** [comment]
I believe WebView2 in WinUI will only work on Desktop. @jevansaks may be able to confirm.

**Comment by alexispolakMURAL** [comment]
By Desktop you mean that it won't work in an app running on Windows teams ?

**Comment by asklar** [comment]
Yes, that is my understanding.

**Comment by jevansaks** [comment]
The latest Windows Teams OS is desktop-based though, right?

**Comment by asklar** [comment]
if that's the case it sounds like this should work if they can install the webview2 runtime, but how do they install the runtime in these devices? maybe a question for @champnic

**Comment by alexispolakMURAL** [comment]
> The latest Windows Teams OS is desktop-based though, right?

@jevansaks from which version is Windows Teams OS is desktop-based ? 

> if that's the case it sounds like this should work if they can install the webview2 runtime, but how do they install the runtime in these devices? maybe a question for @champnic

@asklar that's my main question. I would like to be able to include the WebView2 runtime binaries inside my application so that an external installation is not necessary, but idk if that's possible because the documentation is not clear about it.

**Comment by champnic** [comment]
@alexispolakMURAL I'll open an item on WebView2 to track deployment of the runtime binaries through the store to support Teams. I believe we may have some ongoing work with MSIX that would solve this.

In the meantime, step 3 of the doc means you need to deploy the folder containing the unpacked binaries along with your app somewhere (note: this will make you app install much larger). Step 4 means you will need to point the WebView2 to use those binaries. Unfortunately there isn't a super-straightforward way to do this in the WinUI WebView2 controls, but you should be able to set the `WEBVIEW2_BROWSER_EXECUTABLE_FOLDER` environment variable to the path to the folder containing the unpacked binaries. An additional step for UWP is to make sure your app has permissions to that folder, so it probably needs to live in your app folder.

I'm opening a bug on our end to document how to use Fixed Version binaries in UWP/WinUI. Let me know if you have more questions. Thanks!

**Comment by alexispolakMURAL** [comment]
@champnic thanks ! Can I have a link to the issue ticket ?

**Comment by champnic** [comment]
@alexispolakMURAL https://github.com/MicrosoftEdge/WebView2Feedback/issues/1851

**Comment by nickprotop** [comment]
[https://github.com/microsoft/microsoft-ui-xaml/issues/6150](https://github.com/microsoft/microsoft-ui-xaml/issues/6150) Pretty much the same feature request. The solution with the PATH is working with UWP applications?

**Comment by nickprotop** [comment]
Found a solution. At least for WinUI 3 with Project Reunion 0.8 version.
I have added the fixed distribution of WebView2 into package project within the webview2_64 folder. Then in app.cs:

```
StorageFolder localFolder = Windows.ApplicationModel.Package.Current.InstalledLocation;
Environment.SetEnvironmentVariable("WEBVIEW2_BROWSER_EXECUTABLE_FOLDER", Path.Combine(localFolder.Path, "webview2_64"));
```

And before using WebView2 Control:

`await WebView.EnsureCoreWebView2Async();`

Thanks @champnic for the tip

**Comment by alexispolakMURAL** [comment]
I end up doing what @nickprotop suggested and got the WinUI2 app running with WebView2 Fixed Runtime but got a lot of errors with some `.dll` files when I uploaded the app to the store.

`API PathFindNextComponentW in shlwapi.dll is not supported for this application type. onnxruntime.dll calls this API.
API PathRemoveBackslashW in shlwapi.dll is not supported for this application type. onnxruntime.dll calls this API.
API PathSkipRootW in shlwapi.dll is not supported for this application type. onnxruntime.dll calls this API.
API CoRegisterInitializeSpy in ole32.dll is not supported for this application type. notification_helper.exe calls this API.
API CoRevokeInitializeSpy in ole32.dll is not supported for this application type. notification_helper.exe calls this API.
API BuildExplicitAccessWithNameW in advapi32.dll is not supported for this application type. notification_helper.exe calls this API.
API BuildSecurityDescriptorW in advapi32.dll is not supported for this application type. notification_helper.exe calls this API.
API SHGetFolderPathW in shell32.dll is not supported for this application type. notification_helper.exe calls this API.
API SHGetKnownFolderPath in shell32.dll is not supported for this application type. notification_helper.exe calls this API.
API ShellExecuteExW in shell32.dll is not supported for this application type. notification_helper.exe calls this API.
API AllowSetForegroundWindow in user32.dll is not supported for this application type. notification_helper.exe calls this API.
API SendInput in user32.dll is not supported for this application type. notification_helper.exe calls this API.
API PathMatchSpecW in shlwapi.dll is not supported for this application type. notification_helper.exe calls this API.
API GetFileVersionInfoSizeW in version.dll is not supported for this application type. notification_helper.exe calls this API.
API CreateCompatibleDC in gdi32.dll is not supported for this application type. libGLESv2.dll calls this API.
API CreateDIBSection in gdi32.dll is not supported for this application type. libGLESv2.dll calls this API.
API DeleteDC in gdi32.dll is not supported for this application type. libGLESv2.dll calls this API.
API DeleteObject in gdi32.dll is not supported for this application type. libGLESv2.dll calls this API.`

The list is biger than this one, but didn't want to make the comment super big
